Reworked netbox_ip_address

FragmentedPacket opened the pull request:
Reworked the module to be more inline with the outstanding PRs for netbox_device.py and netbox_prefix.py. This is due to some changes with the pynetbox API and how it returns `create` data.

Added some further if statements when VRFs do not exist with more detail as to the failure.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
netbox_ip_address

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
netbox_device.py rework: https://github.com/ansible/ansible/pull/50672
This module is re-worked similar to the above PR, but returns `ip_address` instead of `device` and is the change in return is documented within this PR.

FragmentedPacket committed:
Reworked netbox_ip_address

Reworked the module to be more inline with the outstanding PRs for netbox_device.py and netbox_prefix.py.

Added some further if statements when VRFs do not exist with more detail as to the failure.

FragmentedPacket commented:
@Anthony25 Here is the updated netbox_ip_address module. Let me know if/what changes you'd like to make as well.

Anthony25 commented:
Thanks a lot, I finally didn't have time to do it yesterday. I'm reviewing your changes ;)

aruhier commented on lib/ansible/module_utils/net_tools/netbox/netbox_utils.py:
```diff
@@ -128,7 +128,10 @@ def find_ids(nb, data):
+                    if vrf_id:
+                        query_id = nb_endpoint.get(**{"address": v["address"], "vrf_id": vrf_id.id})
+                    else:
+                        return "%s does not exist - Please create VRF" % (data["vrf"])
```
It's also valid for the next block, but why don't you raise an exception like `ValueError`, with your message, that you catch when you call `find_ids`?

For example:

```
if vrf_id:
    query_id = nb_endpoint.get(address=v["address"], vrf_id=vrf_id.id)
else:
    raise ValueError("%s does not exist - Please create VRF" % (data["vrf"]))
```

And then in main function:

```

    try:
        if 'present' in state:
            return module.exit_json(
                **ensure_ip_address_present(nb, nb_endpoint, norm_data)
            )
        else:
            return module.exit_json(
                **ensure_ip_address_absent(nb, nb_endpoint, norm_data)
            )
    except pynetbox.RequestError as e:
        return module.fail_json(msg=json.loads(e.error))
    except ValueError as e:
        return module.fail_json(msg=str(e))
```

This way, it's more obvious than there is an error here. You can also define a custom exception for not found objects.

FragmentedPacket replied:
The one thing I'm battling with is if we'd want it to fail closed or open. If we fail closed, that means someone can't react to the `msg` and say create the VRF the next play with a `when` condition.

What are your thoughts?

aruhier replied:
Hmm sorry, I don't really get your question. Could you give me an example please?

FragmentedPacket replied:
Ignore me! After thinking about this further - it makes sense to just have the playbook fail if one of the values doesn't return an id (doesn't exist)!

FragmentedPacket replied:
How does everything look now?

I did notice we're returning a dictionary when we encounter the following which is not what we're documenting in the modules:
```
except pynetbox.RequestError as e:
        return module.fail_json(msg=json.loads(e.error))
```
If we change it to:
```
except pynetbox.RequestError as e:
        return module.fail_json(msg=e.error)
```
It returns the following:
```
TASK [Create prefix with several specified options] *********************************************************************************************************************
fatal: [localhost]: FAILED! => changed=false
  msg: '{"vlan":["Primary key must be an integer"]}'
```

aruhier replied:
LGTM

Hmm, about the error, it's kind of tricky here… I would let it as it is, because I cannot really find a good way to do that :(

FragmentedPacket replied:
Ok awesome! Thanks for reviewing.

I'm going to try and possibly get an `netbox_interface` and `netbox_vrf` one created within the next couple of weeks.

FragmentedPacket committed:
Added raise ValueError to netbox_utils and netbox_ip_address to properly handle failure scenarios.

Anthony25 commented:
I thought about one feature for this module: we could add the possibility to query a new address if no address is specified in the data, and a prefix id is given, using the endpoint `/api/ipam/prefixes/{id}/available-ips/`.

If you are interested about this feature, I can work on it tomorrow.

FragmentedPacket commented:
@Anthony25 Do you mind showing me what that would look like? I was thinking of adding something like that down the road in regards to both IP's and prefixes, but just wanted to get some modules implemented.

Commit:
netbox_ip_address: Refactor to share more code

Anthony25 commented:
@FragmentedPacket: sure, I started my work here: https://github.com/Anthony25/ansible/tree/netbox_ip_address_return_fix

I don't really know how I can contribute to this PR without write-access on your repository, so we will see when I will finish to work on this (I hope in a few hours).

I thought at doing something like this:

Case with interface specified:
========================

```
      netbox_ip_address:
        netbox_url: http://netbox.local
        netbox_token: thisIsMyToken
        data:
          prefix: 192.168.1.0/24
          vrf: Test
          interface:
            name: GigabitEthernet1
            device: test100
        state: present
```

What would check if an IP is already present on test100, interface GigabitEthernet1, in the prefix 192.168.1.0/24 vrf Test. If not, it will try to allocate a new IP in this prefix, and attach it to this interface.

Case with interface specified and state "new":
====================================

```
      netbox_ip_address:
        netbox_url: http://netbox.local
        netbox_token: thisIsMyToken
        data:
          prefix: 192.168.1.0/24
          vrf: Test
          interface:
            name: GigabitEthernet1
            device: test100
        state: new
```

Allocate a new IP for this interface, even if one in the same prefix is already attached.

Case without an interface specified:
============================

```
      netbox_ip_address:
        netbox_url: http://netbox.local
        netbox_token: thisIsMyToken
        data:
          prefix: 192.168.1.0/24
          vrf: Test
        state: new
```

What would return a new IP in 192.168.1.0/24

Case without an interface specified and state "present":
============================================

```
      netbox_ip_address:
        netbox_url: http://netbox.local
        netbox_token: thisIsMyToken
        data:
          prefix: 192.168.1.0/24
          vrf: Test
        state: present
```

Here, I don't really know what to do. To me it would be an error, because "ip present in 192.168.1.0/24" doesn't make a lot of sense and would be a weird case.

What do you think about it?

Commit:
netbox_ip_address: add a state "new"

Allows to force the creation of a new address, even if it already exists
(useful for anycast or for vrf allowing dupplicates). If no address is
specified, will ask for a new ip inside a specific prefix.

TODO: add documentation for it

Anthony25 commented:
I also added one case:

## Case with an address specificied, state "new":
```
      netbox_ip_address:
        netbox_url: http://netbox.local
        netbox_token: thisIsMyToken
        data:
          address: 192.168.1.1/24
          vrf: Test
          role: Anycast
        state: new
```

Will force to create a new address `192.168.1.1/24` even if it already exists. Could be useful for anycasted addresses, or when vrf allow duplicates.

Commit:
netbox_ip_address: get available IP

Ensure an IP in a specified prefix is attached on an interface

Anthony25 commented:
I added those cases in my branch. I just need to test it and complete the documentation, but you can already review if you want to: https://github.com/Anthony25/ansible/commits/netbox_ip_address_return_fix

Commit:
netbox_ip_address: bugfixing

Commit:
netbox_ip_address: add doc

Commit:
Add first_available parameter

Commit:
Revert "Add first_available parameter"

This reverts commit b040b0819d35fed1b99239f1a59837bbc182ecd1.

Anthony25 commented:
@FragmentedPacket: could you tell me if it's okay for you, and if it is could you pull it into your branch please?

Anthony25 commented:
Thanks a lot ;)

FragmentedPacket committed:
Tested use cases and everything appears to work as intended

FragmentedPacket committed:
Fixed conflicts with devel

FragmentedPacket committed:
Fixed conflicts with devel - attempt #2

FragmentedPacket committed:
Merge branch 'devel' into netbox_ip_address_return_fix

FragmentedPacket committed:
Rebased to devel and then added changes of rework

FragmentedPacket committed:
Fixed PEP issues

FragmentedPacket commented:
Sorry for all the messy attempts at merging our changes.

Created a new PR to just have a clean history: #52415.

Closing this one.
